### PR TITLE
generate unique mapId for new entities

### LINKED ISSUE #176: Generate mapIds for EventTrigger
The game uses these in order to determine whether or not it has been triggered.

### PR DESCRIPTION
**Pull request by Vegita2**
fixes #176

### CODE REVIEW

File: webapp/src/app/shared/phaser/entities/cc-entity.ts [hunk 1]
```diff
@@ -31,6 +31,12 @@ export interface AttributeValue {
 	[key: string]: any;
 }
 
+export interface DetailSettings {
+	mapId: number;
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/entities/cc-entity.ts:35]: All uses of `mapId` act as if it were optional. One even does `Partial<DetailSettings>` just for that.

File: webapp/src/app/shared/phaser/entities/cc-entity.ts [hunk 2]
```diff
@@ -95,7 +107,7 @@ export abstract class CCEntity extends BaseObject {
 		pivot: Point;
 	} = <any>{};
 	
-	protected constructor(scene: Phaser.Scene, map: CCMap, x: number, y: number, typeName: string) {
+	public constructor(scene: Phaser.Scene, map: CCMap, x: number, y: number, typeName: string) {
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/entities/cc-entity.ts:110]: Why was this changed? Seeing the commet of `getEntity` this might be a leftover?

File: webapp/src/app/shared/phaser/entities/entity-manager.ts [hunk 3]
```diff
@@ -296,12 +283,36 @@ export class EntityManager extends BaseObject {
...
+		}, true);
+		
+		return e;
+	}
+	
 	async generateEntity(entity: MapEntity): Promise<CCEntity> {
 		const entityClass = Globals.entityRegistry.getEntity(entity.type);
-		
-		const ccEntity = new entityClass(this.scene, this.map, entity.x, entity.y, entity.type);
+		console.assert(this.map);
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/entities/entity-manager.ts:308]: Add description to assert

File: webapp/src/app/shared/phaser/entities/registry/entity-registry.service.ts [hunk 4]
```diff
@@ -1,45 +1,43 @@
...
+	private defaultEntity = DefaultEntity;
 	
 	constructor() {
-		this.setDefaultEntity(DefaultEntity);
 		this.register('Prop', Prop);
 		this.register('ScalableProp', ScalableProp);
 		this.register('ItemDestruct', ItemDestruct);
 		this.register('Destructible', Destructible);
 		this.register('NPC', NPC);
+		this.register('NPC', NPC);
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/entities/registry/entity-registry.service.ts:22]: This looks like it was added by accident

File: webapp/src/app/shared/phaser/entities/registry/entity-registry.service.ts [hunk 5]
```diff
@@ -1,45 +1,43 @@
...
 	public getAll() {
 		return this.entities;
 	}
 	
-	public getEntity(type: string): new (...args: any[]) => CCEntity {
-		return this.entities[type] || this.defaultEntity;
+	// typed as DefaultEntity so constructor can be used with parameter checking.
+	// CCEntity is abstract and doesn't allow using the constructor
+	public getEntity(type: string): typeof DefaultEntity {
+		return (this.entities[type] || this.defaultEntity) as any;
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/entities/registry/entity-registry.service.ts:41]: Why not use `DefaultEntity` directly? 
```suggestion
		return (this.entities[type] ?? DefaultEntity) as typeof DefaultEntity
```
(I don't care of one does `??` or `||` in this case, just a suggestion)

File: webapp/src/app/shared/phaser/tilemap/cc-map.ts [hunk 6]
```diff
@@ -88,8 +89,16 @@ export class CCMap {
 			this.inputLayers = undefined;
 		}
 		
+		this.lastMapId = 0;
```
Comment (2767mr) [inline webapp/src/app/shared/phaser/tilemap/cc-map.ts:92]: This is set to `0` while it is set to `1` at the declaration above

**Review by 2767mr — CHANGES_REQUESTED** [review]
If entities are copied, the `mapId` is also copied.

File: webapp/src/app/shared/phaser/entities/cc-entity.ts (hunk 1 above)
Comment (Vegita2) [reply webapp/src/app/shared/phaser/entities/cc-entity.ts:35]: I moved the call to `map.getUniqueMapid()` to better reflect what i meant with that. It basically says imported map can contain entities without a mapId, but the internal representation of entities always has a mapId.

It is currently unused but having the type is still better than using `any`

File: webapp/src/app/tests/test-helper.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+		return new Promise(((res) => {
+			let imported: CrossCodeMap;
+			let sub: Subscription;
+			// eslint-disable-next-line prefer-const
+			sub = service.tileMap.subscribe(map => {
+				if (!map) {
+					return;
+				}
+				
+				sub!.unsubscribe();
```
Comment (2767mr) [inline webapp/src/app/tests/test-helper.ts:19]: There seems to be a some sort of race condition or test ordering issue here because one test failed because `sub` was undefined. Since I couldn't reprocude the issue I suggest making a note/issue and investigating later.

**Review by 2767mr — APPROVED** [review]
